codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/XmlShapeDeserVisitor.java

```diff
@@ -404,10 +406,67 @@ protected void deserializeDocument(GenerationContext context, DocumentShape shap
     @Override
     protected void deserializeUnion(GenerationContext context, UnionShape shape) {
         GoWriter writer = context.getWriter();
-        // TODO: implement union deserialization
-        // The tricky part is going to be accumulating bytes for unknown members.
-        LOGGER.warning("Union type is currently unsupported for XML deserialization.");
-        context.getWriter().writeDocs("TODO: implement union serialization.");
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        Symbol symbol = symbolProvider.toSymbol(shape);
+        Model model = context.getModel();
+
+        writer.write("var uv $P", symbol);
+        writer.write("var memberFound bool");
+
+        // Iterate through the decoder. The member visitor will handle popping
+        // xml tokens enclosed within a xml start and end element.
+        writer.openBlock("for {", "}", () -> {
+            writer.write("t, done, err := decoder.Token()");
+            writer.write("if err != nil { return err }");
+            writer.write("if done { break }");
+            writer.openBlock("if memberFound {", "}", () -> {
+                writer.write("if err = decoder.Decoder.Skip(); err != nil { return err }");
+            });
+
+            // Create a new decoder for each member
+            writer.write("originalDecoder := decoder");
+            writer.write("decoder = smithyxml.WrapNodeDecoder(originalDecoder.Decoder, t)");
+            writer.insertTrailingNewline();
+
+            writer.openBlock("switch {", "}", () -> {
+                Set<MemberShape> members = new TreeSet<>(shape.members());
+                for (MemberShape member : members) {
+                    // check if member is not a document binding or has a xmlAttribute trait
+                    if (!memberFilter.test(member) || member.hasTrait(XmlAttributeTrait.ID)) {
+                        continue;
+                    }
+                    Symbol targetSymbol = symbolProvider.toSymbol(member);
+                    Symbol memberSymbol = SymbolUtils.createValueSymbolBuilder(
+                            symbolProvider.toMemberName(member),
+                            symbol.getNamespace()
+                    ).build();
+
+                    writer.addUseImports(SmithyGoDependency.STRINGS);
+
+                    String serializedMemberName = getSerializedMemberName(member);
+                    writer.openBlock("case strings.EqualFold($S, t.Name.Local):", "", serializedMemberName, () -> {
+                        writer.write("var mv $P", targetSymbol);
+                        model.expectShape(member.getTarget()).accept(getMemberDeserVisitor(member, "mv", false));
+                        writer.write("uv = &$T{Value: mv}", memberSymbol);
+                    });
+                    writer.write("memberFound = true");
```

Comment:
minor nit: Probably should be in the `openBlock` above since this is the logical place this statement is being executed.